dot/simswap/util/reverse2original.py

```diff
@@ -100,89 +106,103 @@ def reverse2wholeimage(
+
+        orisize = (oriimg.shape[0], oriimg.shape[1])
         if use_mask:
             source_img_norm = norm(source_img, use_gpu=use_gpu)
             source_img_512 = F.interpolate(source_img_norm, size=(512, 512))
             out = pasring_model(source_img_512)[0]
-            parsing = out.squeeze(0).detach().cpu().numpy().argmax(0)
-            vis_parsing_anno = parsing.copy().astype(np.uint8)
-            tgt_mask = encode_segmentation_rgb(vis_parsing_anno)
+            parsing = out.squeeze(0).argmax(0)
+
+            tgt_mask = encode_segmentation_rgb(parsing, device)
+
+            # If the mask is large
             if tgt_mask.sum() >= 5000:
-                target_mask = cv2.resize(tgt_mask, (crop_size, crop_size))
+
+                target_mask = ko_transform.resize(tgt_mask, (crop_size, crop_size))
+
                 target_image_parsing = postprocess(
                     swaped_img,
-                    source_img[0].cpu().detach().numpy().transpose((1, 2, 0)),
+                    source_img[0],
                     target_mask,
                     smooth_mask,
-                    use_gpu=use_gpu,
+                    device=device,
                 )
 
-                target_image = cv2.warpAffine(target_image_parsing, mat_rev, orisize)
+                target_image_parsing = target_image_parsing[None, ...]
+                swaped_img = swaped_img[None, ...]
+
+                target_image = ko_transform.warp_affine(
+                    target_image_parsing, mat_rev, orisize
+                )
             else:
-                target_image = cv2.warpAffine(swaped_img, mat_rev, orisize)[..., ::-1]
+                swaped_img = swaped_img[None, ...]
+                target_image = ko_transform.warp_affine(
+                    swaped_img,
+                    mat_rev,
+                    orisize,
+                )
         else:
-            target_image = cv2.warpAffine(swaped_img, mat_rev, orisize)
-
-        img_white = cv2.warpAffine(img_white, mat_rev, orisize)
-
-        img_white[img_white > 20] = 255
-
-        img_mask = img_white
-
-        kernel = np.ones((40, 40), np.uint8)
-        img_mask = cv2.erode(img_mask, kernel, iterations=1)
-        kernel_size = (20, 20)
-        blur_size = tuple(2 * i + 1 for i in kernel_size)
-        img_mask = cv2.GaussianBlur(img_mask, blur_size, 0)
+            swaped_img = swaped_img[None, ...]
+            target_image = ko_transform.warp_affine(
+                swaped_img,
+                mat_rev,
+                orisize,
+            )
 
-        img_mask /= 255
+        img_white = ko_transform.warp_affine(img_white, mat_rev, orisize)
 
-        img_mask = np.reshape(img_mask, [img_mask.shape[0], img_mask.shape[1], 1])
+        img_white[img_white > 0.0784] = 1.0
 
-        if use_mask:
-            target_image = np.array(target_image, dtype=np.float) * 255
+        if use_cam:
+            kernel = torch.ones(5, 5).to(device)
```

Comment:
move out of the loop

Reply:
same as any constant tensor in the `else`

Reply:
Done https://github.com/sensity-ai/dot/pull/20/commits/40051c2389b20d24592099882ebe80fa808a7428